### ISSUE TITLE
Improving handling of missing values

### ISSUE THREAD

**Issue by aarmey** [tensorly/tensorly#198]
Currently, we perform masking to fill in missing values at each iteration. While this works, for increasing amounts of missingness (~10%+) it greatly slows convergence. ([Some experiments showing this here.](http://www.models.life.ku.dk/sites/default/files/2005,%20PARAFAC%20missing%20values.pdf)) I'd be interested in exploring ways of improving this as I've been working with data having ~25-30% missing values. Two possibilities:
- I'm aware of two papers that propose solutions to this, both using generic optimization with weighted gradient calculations to only consider non-missing values ([1](http://www.models.life.ku.dk/sites/default/files/2005,%20PARAFAC%20missing%20values.pdf), [2](http://www.cs.sandia.gov/~dmdunla/publications/AcDuKoMo10.pdf)). 
- Outside of tensor methods, I know there are methods for setting up [least squares solving with missing values](http://alexhwilliams.info/itsneuronalblog/2018/02/26/censored-lstsq/). I'm not sure if there are reasons this hasn't been using previously with ALS.

I agree with @cohenjer that ALS adjustments, if possible, would be better than using generic optimization.

@cohenjer and @JeanKossaifi, perhaps you have thoughts?

**Comment by JeanKossaifi** [comment]
Better handling of missing values would indeed be useful. 

Another consideration is that to recover missing values, we typically make strong assumptions on the low-rank structure. [Robust decomposition](http://tensorly.org/dev/modules/generated/tensorly.decomposition.robust_pca.html#tensorly.decomposition.robust_pca) tends to shine there by also explicitly modelling gross corruption (I had written a [blog post](http://jeankossaifi.com/blog/rpca.html) on robust tensor PCA in TensorLy on very noisy data). We could expand Robust PCA to this case by either
1. first using and then decomposing the low-rank part or 
2. directly learning a factorized representation of the low-rank part. I had written some code for this but never got to finalizing it, I could try to dig it up if it is of interest.

ALS adjustment would be a neat way to improve the existing algorithm. The computational cost of doing this is typically high though. We could speed things up if we know the missing values are structured  (e.g. if we can sample *fibers* from the tensors (in other words we have missing values on all modes but one), we could accordingly [sample the Khatri-Rao product](https://github.com/tensorly/tensorly/blob/2dbd582272a79b245307d484c17aacf142d5f6a1/tensorly/decomposition/candecomp_parafac.py#L559) as in the randomized CP).

**Comment by cohenjer** [comment]
Hey,
So if I understood correctly, we are talking about missing data in tensors when computing CPD ? I think other models will lead to slightly different discussion, so let me focus on CPD.

@aarmey you are right that there has not been a serious discussion (as far as I know at least) about missing data and ALS apart from imputation techniques (i.e. replacing unknown values with guesses, computing the CPD, and repeating that process until some convergence, which is quite costly). Other algorithms are fine, but at the moment we only have ALS in tensorly so it would be nice to have a better handling of missing values until maybe other algorithms are added later on.

As @JeanKossaifi mentionned, this lack of study is probably due to the fact that for unstructured missing data, the ALS update rules are not as nicely implemented. I was not aware of Alex William's post, but we can actually use it as a basis for a small discussion on how we could implement this. Discussing maths in github comments is horrible however, so I'll just sketch the idea:

Say we are updating factor A in a kruskal tensor [A,B,C]. Let W the mask of missing data (tensor of ones and zeros). Then we are in fact solving a matrix linear system with missing values.
![image](https://user-images.githubusercontent.com/22053798/95201565-6d22d480-07e0-11eb-8e11-209123877af8.png)
We can solve this in closed form row by row, which means we can focus on a simpler problem 
![image](https://user-images.githubusercontent.com/22053798/95201709-96dbfb80-07e0-11eb-965b-8008b6a299eb.png)
 with t, w beings rows of T_{1} and W_{1}. As well described in Alex's post, this is equivalent to a problem without w where we remove columns in y and $D:=(B\odot C)^T$ (i.e. rows in $B\odot C$). Let's note t' and D' these smaller vector/matrix. We end up with computing some MTTKRP $t'D'^T$ for all t rows of $T_{1}$, which has the not so nice property of not being castable as a sequence of multiway products, hence we can't call efficient tensor contraction libraries. Similarly, we compute $D' D'^T$ which is not simplified as $B^TB \ast C^TC$ because we removed some rows in D. So computing $D' D'^T$ will be a little costly, of order O(JK R^2) where J and K are the 2d and 3d dimensions. But it's definitively doable by first computing (B\odot C) (outside the loop on t) then removing rows, and performing a matrix-matrix product.

Bottom line, this is doable; this would require some extended version of this sketch for any user to understand. The algorithm will run a little bit slower but I believe that for dense tensors, there should not be a huge difference with regular ALS. But this strategy is also not studied currently so we can't really say yet if it's worth the trouble.

Final word: if the missing data are structured, we can definitively simplify all this. However, the structure we need would be on all three modes; that is to say, W should be a rank-one tensor which I think is a very strong assumption. If is it structured such that we select fibers along one mode, then we can speed things up when updating the factor on that mode, but the other factors will still be updated with the strategy I mentioned above.

**Comment by cohenjer** [comment]
As a complement, the best discussion on missing data and ALS is probably still Rasmus Bro's thesis, pages 146-147.

**Comment by aarmey** [comment]
Just to follow up on this discussion, exploiting the pattern of missing values was *exactly* what I needed. A generic optimization function worked well and was faster than an unmodified ALS, but then modified ALS was easily orders of magnitude faster. Because of the structure of my missing values, I could remove rows after Khatri Rao and the tensor unfolding ([code here](https://github.com/meyer-lab/systemsSerology/blob/master/syserol/tensor.py)).

As for Tensorly, we would build an option to allow this type of handling? If there are missing values above a certain percentage along a fiber we would remove it during the ALS operation. Unlike a generic optimization method this would not introduce additional external dependencies and should work with the various backends. Indeed, it would come with a performance cost but need only be enabled if there is sufficient missingness to make the unmodified ALS convergence too slow.

I see value in making this available to folks who don't want to entirely reimplement ALS themselves, but would add ever more complexity to the ALS code.

**Comment by JeanKossaifi** [comment]
I like your approach @aarmey, the resulting code is very clean! I think we could even directly sample only the part of the khatri-rao that you need (e.g. through [`sample_khatri_rao`](https://github.com/tensorly/tensorly/blob/2dbd582272a79b245307d484c17aacf142d5f6a1/tensorly/decomposition/candecomp_parafac.py#L559)).

The API is the trickiest thing to design for this: we need to now be able to specify either missing values on the whole tensor or per mode. One way could be to have the mask parameter be either a single tensor (for the whole reconstruction) or a list of tuples [(mask, mode)] (or a dict {mode: mask}).

**Comment by aarmey** [comment]
Great. For the API, the simplest would be to have a list of vectors, but I like your suggestion as it leaves open other mode-specific strategies in the future. We can just allow either of the formats you suggest, since it's easy to check and convert one into the other.

I'll setup a PR when I get a chance.

**Comment by cohenjer** [comment]
I also think it would helpful if you could describe precisely what kind of structure for the missing data you can deal with. I guess that you have whole rows/columns/fibers of data missing? If you can remove rows after the unfolding, then could you not remove all the missing rows/columns/fibers from the start?

**Comment by aarmey** [comment]
This would be specifically for fibers—you're right @cohenjer that if a whole slice is missing you should just remove it. Because it is for fibers this can only be done after unfolding.

Your question made me realize that we can't simply ask the user to define the missing entries corresponding to indices for each dimension of the tensor. The entries to be removed are only evident in unfolded form, and I'm not sure we can expect the user to know the structure of unfolding. I suppose we could take a scalar argument like `fiber_mask_pct`, and then remove fibers if they are masked above that threshold?

**Comment by aarmey** [comment]
I now have a much better understanding of this issue and agree there's not an easy general solution.
